qualtran/bloqs/arithmetic/addition.py

```diff
@@ -137,10 +148,18 @@ def wire_symbol(self, soq: 'Soquet') -> 'WireSymbol':
             raise ValueError()
 
     def _left_building_block(self, inp, out, anc, depth):
-        if depth == self.dtype.bitsize - 1:
+        if depth == self.b_dtype.bitsize - 1:
             return
         else:
-            yield CNOT().on(anc[depth - 1], inp[depth])
+            if inp[depth] is not None:
+                yield CNOT().on(anc[depth - 1], inp[depth])
+            else:
+                # If inp[depth] = None, we treat it as a |0>,
+                # and therefore applying CNOT().on(anc[depth - 1], inp[depth])
+                # essentially "copies" anc[depth - 1] into inp[depth]
+                # in the classical basis. So therefore, on future operations,
+                # we can use anc[depth - 1] in its place.
+                inp[depth] = anc[depth - 1]
```

Comment:
this statement is potentially dangerous if the caller assumes that `_left_building_block` doesn't modify its input which is reasonable since its current implementation doesn't.

consider using a a variable instead

```py3

if ...:
   ...
   control = inp[depth]
else:
   control = anc[depth - 1]
```

This will make the logic clearer and remove this condition from below
```py3
            if anc[depth - 1] != inp[depth]:
```